great_expectations/expectations/metrics/multicolumn_map_metrics/compound_columns_unique.py

```diff
@@ -21,16 +35,145 @@ class CompoundColumnsUnique(MulticolumnMapMetricProvider):
+        column_names = kwargs.get("_column_names")
+
+        # Need all columns of the table for the purposes of reporting entire rows satisfying unexpected condition logic.
+        table_columns = kwargs.get("_table_columns")
+
+        table = kwargs.get("_table")
+
+        # First, obtain the SQLAlchemy "FromClause" version of the original "table" for the purposes of gaining the
+        # "FromClause.c" attribute, which is a namespace of all the columns contained within the "FROM" clause (these
+        # elements are themselves subclasses of the SQLAlchemy "ColumnElement" class).
+        table_columns_selector = [
+            sa.column(column_name) for column_name in table_columns
+        ]
+        original_table_query = (
+            sa.select(table_columns_selector).select_from(table).alias("original_table")
+        )
+
+        # Second, "SELECT FROM" the original table, represented by the "FromClause" object, querying all columns of the
+        # table and the count of occurrences of distinct "compound" (i.e., group, as specified by "column_list") values.
+        # Give this aggregated group count a distinctive label.
+        # Give the resulting sub-query a unique alias in order to disambiguate column names in subsequent queries.
+        count_selector = column_list + [sa.func.count().label("_num_rows")]
+        group_count_query = (
+            sa.select(count_selector)
+            .group_by(*column_list)
+            .select_from(original_table_query)
+            .alias("group_counts")
+        )
+
+        # The above "group_count_query", if executed, will produce the result set containing the number of rows that
+        # equals the number of distinct values of the group -- unique grouping (e.g., as in a multi-column primary key).
+        # Hence, in order for the "_num_rows" column values to provide an entry for each row of the original table, the
+        # "SELECT FROM" of "group_count_query" must undergo an "INNER JOIN" operation with the "original_table_query"
+        # object, whereby all table columns in the two "FromClause" objects must match, respectively, as the conditions.
+        conditions = sa.and_(
+            *(
+                group_count_query.c[name] == original_table_query.c[name]
+                for name in column_names
+            )
+        )
+        # noinspection PyProtectedMember
+        compound_columns_count_query = (
+            sa.select(
+                [original_table_query, group_count_query.c._num_rows.label("_num_rows")]
+            )
+            .select_from(
+                original_table_query.join(
+                    right=group_count_query, onclause=conditions, isouter=False
+                )
+            )
+            .alias("records_with_grouped_column_counts")
+        )
+
+        # The returned SQLAlchemy "FromClause" "compound_columns_count_query" object realizes the "map" metric function.
+        return compound_columns_count_query
```

Comment:
🥳 This seems like the most computationally efficient way using SQL. I would be curious to see how the `group_by` performs on large datasets vs Pandas (Spark should absolutely outperform on a large dataset).

Reply:
@NathanFarmer In my opinion, this can be a "canonical" pattern for us going forward where we need to implement a `self-join` as part of computing the unexpected condition for the "Map" style metrics.  It is extremely efficient, because all computations are done in `SQL` and are executed only at the end, when the entire query is sent to the database engine.  Thanks!